File: go/internal/audit/report.go
package audit

import (
	"context"
	"fmt"
	"path"
	"slices"
	"time"

	"github.com/chadit/LinodeMCP/go/internal/config"
	"github.com/chadit/LinodeMCP/go/internal/genlocal"
)

// One named report, run over the same stores the other query operations read.
//
// The engine sits here rather than beside the tool because every decision it
// makes belongs to the audit subsystem: which store answers, what a window
// means, which events a filter admits, and what a definition the vocabulary
// will not take does. The tool layer is left with the two conditions those
// decisions report.

// definitionError marks one failure as the definition's own while leaving its
// sentence alone, because the sentence is what a caller reads back and a wrap
// would double the condition into it.
type definitionError struct{ cause error }

func (d definitionError) Error() string { return d.cause.Error() }

// Unwrap answers the condition beside the cause, so errors.Is finds
// ErrReportDefinition without the sentence carrying its words.
func (d definitionError) Unwrap() []error { return []error{ErrReportDefinition, d.cause} }

// Report runs one named report over the configured database where one answers
// and over the JSONL log otherwise.
//
// It reports ErrReportDefinition for a definition the vocabulary will not take
// and answers whatever the read failed with otherwise. A configured database
// that will not open leaves the log standing behind a warning, which is the
// store's own degrade rule.
//
// now is the reference instant a relative window is measured back from. It is
// handed in rather than read here so a pinned clock reaches the window, which
// is what lets a report fixture replay.
func Report(
	ctx context.Context, sqlitePath, jsonlDir, name string,
	definition *config.ReportConfig, now time.Time,
) (*genlocal.AuditReportResponse, error) {
	query, err := reportQuery(&definition.Filter, now)
	if err != nil {
		return nil, definitionError{cause: err}
	}

	events, warnings, err := ReadWithFallback(sqlitePath,
		func(store string) ([]*Event, error) {
			return ExportEvents(ctx, store, jsonlDir, query)
		})
	if err != nil {
		return nil, err
	}

	return reportAnswer(name, definition, matchingReportEvents(events, &definition.Filter), warnings)
}

// reportAnswer is the report's own answer, built once with its warnings in
// whichever output the definition names.
//
// Both outputs fill both list members because the answer declares both and the
// one the output did not build is answered empty rather than left out.
func reportAnswer(
	name string, definition *config.ReportConfig, matched []*Event, warnings []string,
) (*genlocal.AuditReportResponse, error) {
	if definition.Output == config.ReportOutputSummary {
		columns, err := ValidateGroupBy(definition.GroupBy)
		if err != nil {
			return nil, definitionError{
				cause: fmt.Errorf("validate report group_by: %w", err),
			}
		}

		return genlocal.NewAuditReportResponse(name, config.ReportOutputSummary,
			len(matched), Summarize(matched, columns), []*Event{}, warnings), nil
	}

	if definition.Limit > 0 && len(matched) > definition.Limit {
		matched = matched[:definition.Limit]
	}

	return genlocal.NewAuditReportResponse(name, definition.Output, len(matched),
		[]*genlocal.AuditSummaryRow{}, matched, warnings), nil
}

// reportQuery translates the filter members a store query carries into one
// query.
//
// IncludeMeta is true because the report grammar decides meta inclusion through
// its own capability filter rather than through the tool-layer default, and
// since_offset wins over the absolute since when both are set.
func reportQuery(filter *config.ReportFilter, now time.Time) (*RecentQuery, error) {
	since, err := reportSince(filter, now)
	if err != nil {
		return nil, err
	}

	until, err := reportBound(filter.Until, "until")
	if err != nil {
		return nil, err
	}

	return &RecentQuery{
		Limit:       MaxExportRecords,
		Since:       since,
		Until:       until,
		Tool:        filter.Tool,
		Capability:  Capability(filter.Capability),
		Status:      Status(filter.Status),
		IncludeMeta: true,
	}, nil
}

// reportSince is the load-time lower bound: the offset measured back from the
// reference instant where the filter names one, else the absolute timestamp,
// else no bound.
func reportSince(filter *config.ReportFilter, now time.Time) (time.Time, error) {
	if filter.SinceOffset == "" {
		return reportBound(filter.Since, "since")
	}

	span, err := time.ParseDuration(filter.SinceOffset)
	if err != nil {
		return time.Time{}, fmt.Errorf("parse since_offset %q: %w", filter.SinceOffset, err)
	}

	return now.Add(-span), nil
}

// reportBound parses one absolute bound, answering the zero time for an absent
// one. The configuration validator reads these first, so a value that fails
// here reached the operation from a file nothing checked.
func reportBound(value, member string) (time.Time, error) {
	if value == "" {
		return time.Time{}, nil
	}

	parsed, err := time.Parse(time.RFC3339, value)
	if err != nil {
		return time.Time{}, fmt.Errorf("parse %s %q: %w", member, value, err)
	}

	return parsed, nil
}

// matchingReportEvents keeps the events the filter members a store query does
// not carry admit: the two membership lists, and the environment and profile
// globs.
func matchingReportEvents(events []*Event, filter *config.ReportFilter) []*Event {
	matched := make([]*Event, 0, len(events))

	for _, event := range events {
		if reportAdmits(event, filter) {
			matched = append(matched, event)
		}
	}

	return matched
}

// reportAdmits is whether one event clears every post-load filter member.
func reportAdmits(event *Event, filter *config.ReportFilter) bool {
	if len(filter.CapabilityIn) > 0 && !slices.Contains(filter.CapabilityIn, event.ToolCapability) {
		return false
	}

	if len(filter.StatusIn) > 0 && !slices.Contains(filter.StatusIn, event.Status) {
		return false
	}

	return globAdmits(filter.Environment, event.Environment) &&
		globAdmits(filter.Profile, event.Profile)
}

// globAdmits is whether one glob admits a value, with an unset glob admitting
// everything. A malformed pattern matches nothing, which is what path.Match
// answers beside its error.
func globAdmits(pattern, value string) bool {
	if pattern == "" {
		return true
	}

	matched, _ := path.Match(pattern, value)

	return matched
}

File: python/src/linodemcp/audit/report.py
"""One named report, run over the same stores the other query operations read.

Mirrors ``go/internal/audit/report.go``. The engine sits here rather than
beside the tool because every decision it makes belongs to the audit subsystem:
which store answers, what a window means, which events a filter admits, and
what a definition the vocabulary will not take does. The tool layer is left
with the two conditions those decisions report.
"""

from __future__ import annotations

import fnmatch
from datetime import datetime, timedelta
from typing import TYPE_CHECKING

from linodemcp.audit.export import MAX_EXPORT_RECORDS, export_events
from linodemcp.audit.reader import RecentQuery
from linodemcp.audit.store import read_with_fallback
from linodemcp.audit.summary import (
    UnknownGroupByColumnError,
    summarize,
    validate_group_by,
)
from linodemcp.config import REPORT_OUTPUT_SUMMARY, parse_duration_seconds
from linodemcp.genlocal import AuditReportResponse

if TYPE_CHECKING:
    from linodemcp.audit.event import Event
    from linodemcp.config import ReportConfig, ReportFilter


class ReportDefinitionError(ValueError):
    """Raised for a report the configuration wrote wrong.

    An offset that is not a duration, a bound that is not a timestamp, a
    group_by column outside the summary vocabulary. It is the caller's own
    condition rather than a failed read, and the two reach the tool layer
    through one call, so the failure has to say which it is.
    """


def report(
    sqlite_path: str,
    jsonl_dir: str,
    name: str,
    definition: ReportConfig,
    now: datetime,
) -> AuditReportResponse:
    """Run one named report over the configured database where one answers and
    over the JSONL log otherwise.

    Raises ReportDefinitionError for a definition the vocabulary will not take
    and whatever the read failed with otherwise. A configured database that will
    not open leaves the log standing behind a warning, which is the store's own
    degrade rule.

    ``now`` is the reference instant a relative window is measured back from. It
    is handed in rather than read here so a pinned clock reaches the window,
    which is what lets a report fixture replay.
    """
    query = _report_query(definition.filter, now)

    events, warnings = read_with_fallback(
        sqlite_path,
        lambda store: export_events(store, jsonl_dir, query),
    )

    matched = [event for event in events if _report_admits(event, definition.filter)]

    return _report_answer(name, definition, matched, warnings)


def _report_answer(
    name: str,
    definition: ReportConfig,
    matched: list[Event],
    warnings: list[str],
) -> AuditReportResponse:
    """The report's own answer, built once with its warnings in whichever
    output the definition names.

    Both outputs fill both list members because the answer declares both, and
    the one the output did not build is answered empty rather than left out.
    """
    if definition.output == REPORT_OUTPUT_SUMMARY:
        return AuditReportResponse(
            name=name,
            output=REPORT_OUTPUT_SUMMARY,
            total_events=len(matched),
            rows=summarize(matched, _report_columns(definition)),
            events=[],
            warnings=warnings,
        )

    if definition.limit > 0 and len(matched) > definition.limit:
        matched = matched[: definition.limit]

    return AuditReportResponse(
        name=name,
        output=definition.output,
        total_events=len(matched),
        rows=[],
        events=list(matched),
        warnings=warnings,
    )


def _report_columns(definition: ReportConfig) -> list[str]:
    """The columns a summary report groups by, refusing one the vocabulary lacks.

    The cause is re-raised under the definition's own condition so the sentence
    says which validation failed, the way the other language's does.
    """
    try:
        return validate_group_by(list(definition.group_by))
    except UnknownGroupByColumnError as exc:
        msg = f"validate report group_by: {exc}"
        raise ReportDefinitionError(msg) from exc


def _report_query(filter_: ReportFilter, now: datetime) -> RecentQuery:
    """Translate the filter members a store query carries into one query.

    include_meta is True because the report grammar decides meta inclusion
    through its own capability filter rather than through the tool-layer
    default, and since_offset wins over the absolute since when both are set.
    """
    return RecentQuery(
        limit=MAX_EXPORT_RECORDS,
        since=_report_since(filter_, now),
        until=_report_bound(filter_.until),
        tool=filter_.tool,
        capability=filter_.capability,
        status=filter_.status,
        include_meta=True,
    )


def _report_since(filter_: ReportFilter, now: datetime) -> datetime | None:
    """The load-time lower bound: the offset measured back from the reference
    instant where the filter names one, else the absolute timestamp, else none.
    """
    if not filter_.since_offset:
        return _report_bound(filter_.since)

    try:
        span = parse_duration_seconds(filter_.since_offset)
    except ValueError as exc:
        raise ReportDefinitionError(str(exc)) from exc

    return now - timedelta(seconds=span)


def _report_bound(value: str) -> datetime | None:
    """Parse one absolute bound, answering None for an absent one.

    The configuration validator reads these first, so a value that fails here
    reached the operation from a file nothing checked.
    """
    if not value:
        return None

    try:
        return datetime.fromisoformat(value)
    except ValueError as exc:
        raise ReportDefinitionError(str(exc)) from exc


def _report_admits(event: Event, filter_: ReportFilter) -> bool:
    """Whether one event clears every filter member a store query cannot carry.

    The two membership lists, and the environment and profile globs.
    """
    capability = event.tool_capability
    if filter_.capability_in and capability not in filter_.capability_in:
        return False

    if filter_.status_in and event.status not in filter_.status_in:
        return False

    return _glob_admits(filter_.environment, event.environment) and _glob_admits(
        filter_.profile, event.profile
    )


def _glob_admits(pattern: str, value: str) -> bool:
    """Whether one glob admits a value, with an unset glob admitting every one."""
    return not pattern or fnmatch.fnmatchcase(value, pattern)
